### DISCUSSION TITLE
Product Qty Units and Product Qty Minimum

### DISCUSSION THREAD

**Discussion by papanco** [zencart/zencart#5779] [Ideas]
it would be convenient to set the automatic adjustment of
Product Qty Units and Product Qty Minimum in the cart, similar to the Product Qty Maximum: is it possible to insert it?

**Comment by drbyte** [comment]
Hmmm ... I thought it already did that.

**Comment by papanco** [comment]
> 
> 
> Hmmm ... I thought it already did that.

I tried it today on the latest version but only the warning to manually adjust the quantities appears

**Comment by mc12345678** [comment]
Having added the alert message about the discrepancy for units and minimum values, I'm trying to understand the original statement:
> it would be convenient to set the automatic adjustment of
> Product Qty Units and Product Qty Minimum in the cart, similar to the Product Qty Maximum: is it possible to insert it?

I don't see in processing to the preview where an option to auto set/change the maximum is offered.  I do see in the initial page load that there are defaults set:
````php
  'products_quantity_order_min' => '1',
  'products_quantity_order_units' => '1',
 
  'products_quantity_order_max' => '0',
````
and there is a test when displaying at the minimum and units values:
````php
($pInfo->products_quantity_order_min == 0 ? 1 : $pInfo->products_quantity_order_min)
($pInfo->products_quantity_order_units == 0 ? 1 : $pInfo->products_quantity_order_units)
````
What this does is allow a default of some other value to be applied, provided the default is not 0 then the value provided as the default will be populated on applying a new product or loading a screen of an existing product.

Are you asking that in the alert message that if there is a discrepancy between the units and minimum values that there be something like a clickable link to adjust the units to match the minimum and/or the minimum to match the units?  If so, I don't see a similar capability for the maximum entry. 

So again trying to understand the flow process requested.

**Comment by papanco** [comment]
I enclose examples in order to better explain me: if the cart updates itself automatically with the minimum and multiples that can be ordered, everything is simpler
![example-1](https://user-images.githubusercontent.com/41792764/80736979-2b944c00-8b13-11ea-90b2-1321eaf6b625.jpg)
![example-2](https://user-images.githubusercontent.com/41792764/80736996-3222c380-8b13-11ea-817a-e90363ba8f6b.jpg)
![example-3](https://user-images.githubusercontent.com/41792764/80737016-3a7afe80-8b13-11ea-97ef-7bd4bf38c152.jpg)

**Comment by drbyte** [comment]
@papanco You're right: I'd misunderstood your initial question. 

But I'm undecided about whether it's a good idea to change this.

Just to be clear: you're saying you want to INCREASE the number of items a customer ordered, automagically, right?
Which also means if they ignore the warning message and just blindly go through checkout, they've suddenly spent MORE than they consciously chose to buy, right?
 ... where they've NOT had to actually change their quantity manually before being allowed to checkout, right?

How will this impact (ie: increase) the number of complaints and refund requests, and bank chargebacks a store receives?

(Decreasing is slightly different in that you're never over-charging or shipping more than the customer consciously chose to add to their cart.)

Just asking to be clear that this is your intention.

It'd also help to know exactly what kinds of products you're talking about where this becomes critical.

**Comment by papanco** [comment]
@drbyte the idea is precisely that of an automatic adjustment up or down, always showing a message that warns that the trolley has been adapted to the minimum order or units of quantity of product that can be ordered.
Complaints: I have used a similar method for 16 years and no one has ever complained, since the notice of rectification is clearly visible and before concluding the order there are at least 2 other steps in which the order is summarized.

**Comment by drbyte** [comment]
@mc12345678 see the various `actionXyz()` methods on the `shopping_cart` class. Search for `ERROR_MAXIMUM_QTY` for context.

**Comment by drbyte** [comment]
@papanco Thanks for the context.

Curious: Are you talking about buying clothing? electronics transistors? fish? houses?
And what's the business reason for the minimum quantities? 
(I'm asking because it helps to understand the many use-cases for which people are using these features ... as it helps explain the importance of keeping/updating, or retiring, things over time.)

**Comment by papanco** [comment]
@drbyte commercially it can be used in various sales channels, for example a box contains 200 pieces and you want to sell only complete boxes without having to open them to sell them in bulk;
the minimum order can be linked to the price or the minimum request of the supplier if the product is not already in your warehouse.
What is certain is that the customer hardly ever reads, just click and we must be able to read his mind ...

**Comment by drbyte** [comment]
> What is certain is that the customer hardly ever reads, just click and we must be able to read his mind ...

That's definitely the hard part of commerce!

**Comment by mc12345678** [comment]
(A little late in finishing this as there have been some 6+ messages since I started writing... :) )

Ahh, you were referring to the catalog side not the admin side.  Thank you for that representation.

So, there is the potential for a small problem with the autocorrection of the units/min values, namely when product has attributes and the combination of one or more unique products (product A with attribute combination 1 and product A with attribute combination 2) is permitted to meet that minimum/unit combinations.  In that case, a customer may be wanting to order 2 Matrox G400 32MB with Model Premium and 16 mb memory and 1 Matrox G400 32MB of Model Premium and 8 mb memory... If an auto-correction were applied, then the first choosing would force a value of 3, the second choosing would allow 1, but then the customer would have to reduce the quantity of the first product from 3 to 2...

Alternatively, the Min and the Units could match such that there would not likely be a mismatch, at least when there are no attributes and therefore no variants (at which point assigning an "auto-number" would/could make sense.  But it is the "mix" situation where it becomes "difficult" to auto-adjust.

**Comment by mc12345678** [comment]
I just realized something as I was thinking over this conversation. If in fact the quantity of the product was always adjusted to align with min/units, then it would never be possible to have a combination of variants other than where the number of one variant would be the same as a multiple of the units...
E.g. Units 3
Min 2
Adding 2 to the cart would autocorrect to 3.
Attempting to decrease this from 3 would likely result in it autoreturning to 3 based on rounding. If another variant of the product were added, there would be just 1 of that variant, but may again round since adding something to cart. This time though the total is expected to go in as a factor of the units so adding 2 causes 3, adding 2 additional other variant would cause 3 again to be added. Hmm. Perhaps then the quantity of both could be modified; however, when saving the changes almost have to ask for the class to ignore the discrepancy until they all were proccessed.

**Comment by mc12345678** [comment]
Alternatively, perhaps the response message provided when the item count doesn't match the min/units a link could be given to adjust the most recently added product to fulfill the requirement...  Meaning that in the response message (Yellow area presented) A link would be provided with some words about which product would be adjusted and either to what it would be adjusted or how many would be added.  Clicking on the link in that text would then "autoadjust" the quantity.  

In that way, the concerns expressed about "over purchasing" the item(s) added to the cart.  Though, to also address "early notification" of this min/unit mismatch, if the customer doesn't get sent directly to the cart after adding a product, then they don't know about the mismatch until they choose to visit the cart, though at that point the notification I think would have to do a little more work to determine which "final" product is the one to "auto-adjust" though a more complete solution would be to identify to the customer what grouping of product should be adjusted so they can try to choose a suitable quantity to split among the option(s).
